src/internal/storage/fileset/storage.go

```diff
@@ -243,35 +244,33 @@ func (s *Storage) Concat(ctx context.Context, ids []ID, ttl time.Duration) (*ID,
 		if prim == nil {
 			return nil, errors.Errorf("file set %v is not primitive", id)
 		}
-		additiveIndexes = append(additiveIndexes, prim.Additive)
-		deletiveIndexes = append(deletiveIndexes, prim.Deletive)
+		if prim.Additive != nil {
+			if err := additive.WriteIndex(prim.Additive); err != nil {
+				return nil, err
+			}
+		}
+		if prim.Deletive != nil {
+			if err := deletive.WriteIndex(prim.Deletive); err != nil {
+				return nil, err
+			}
+		}
 		size += prim.SizeBytes
 	}
-	additive, err := s.concat(ctx, additiveIndexes)
+	additiveIdx, err := additive.Close()
 	if err != nil {
 		return nil, err
 	}
-	deletive, err := s.concat(ctx, deletiveIndexes)
+	deletiveIdx, err := deletive.Close()
 	if err != nil {
 		return nil, err
 	}
 	return s.newPrimitive(ctx, &Primitive{
-		Additive:  additive,
-		Deletive:  deletive,
+		Additive:  additiveIdx,
+		Deletive:  deletiveIdx,
 		SizeBytes: size,
 	}, ttl)
 }
 
-func (s *Storage) concat(ctx context.Context, indexes []*index.Index) (*index.Index, error) {
```

Comment:
I think the work being done by this previously was mostly wasted (I feel like this came up in the design doc or discussion). The previous code fed each individual file index entry through the index writer, which clearly satisfies all of our guarantees around batching. However, the incoming file sets are already the results of a compaction step, and so have nice indices, so in a sense the `Merge` here was only having an effect at the boundaries (which is not to say it's trivial, it might be that one index should have been effectively nested at a lower level inside of another one), and of course the new code is O(# file sets) instead of O(# files).

An important change here (if I'm following) is that this is the source of the "unbalanced" shards - rather than building the index up from the file level uniformly, we are adding the top level of each incoming file set at a uniform level, despite it being very likely that they have different depths.

I'd want @brycemcanally to chime in, though :)